pyhoma/models.py

```diff
@@ -153,6 +153,11 @@ def __contains__(self, name: str) -> bool:
     def __getitem__(self, name: str) -> Optional[State]:
         return next((state for state in self._states if state.name == name), None)
 
+    def __setitem__(self, name: str, state: State) -> None:
+        if self.__contains__(state.name):
```

Comment:
That's seems unnatural, a set method must set a value, even if this one already exist.

Reply:
although this ensure that an existing state can't be overwritten

Reply:
But if I want to do it? This is not the classic list behavior. @iMicknl  What do you think?

Reply:
I don't see the scenario where you don't want to overwrite an existing state. If it is a valid scenario, I would do it in a separate function. Otherwise it is confusing if we override default behaviour.

Reply:
fine by me

Reply:
I think though that an existing value should be popped before the append, otherwise calling `device.states[state.name]`  will always refer to the first one in the list, and not the one appended.

Reply:
it could be done with
```python
    def __setitem__(self, name: str, state: State) -> None:
        if self.__contains__(state.name):
            i = self._states.index(self.__getitem__(name))
            self._states.pop(i)
        self._states.append(state)
```

Reply:
Why not simply get the matching state with __getitem__, if None append the state to the list, else update the state value.

Reply:
this works:
```py
    def __setitem__(self, name: str, state: State) -> None:
        s = self.__getitem__(name)
        if s is None:
            self._states.append(state)
        else:
            i = self._states.index(self.__getitem__(name))
            self._states[i] = state
```

Reply:
Almost perfect. Just rewrite the I assignment to self._states.index(s)

Reply:
indeed... again with the Shadocks 😋